lib/spack/spack/solver/asp.py

```diff
@@ -2433,14 +2446,23 @@ def visit(node):
         path = os.path.join(parent_dir, "concretize.lp")
         parse_files([path], visit)
 
+    def add_runtime_virtuals(self):
+        """Explicitly add the virtuals that might be attached dynamically by runtimes"""
+        for vpkg in ("fortran-rt", "ifcore", "gfortran"):
+            self.possible_virtuals.add(vpkg)
+            self.gen.fact(fn.possible_in_link_run(vpkg))
+
     def define_runtime_constraints(self):
         """Define the constraints to be imposed on the runtimes"""
         recorder = RuntimePropertyRecorder(self)
         for compiler in self.possible_compilers:
-            if compiler.name != "gcc":
+            if compiler.name not in ("gcc", "oneapi"):
```

Comment:
We shouldn't special case things here. 

First thought: this should look at whether the compiler's corresponding package class has a `runtime_constraints()` method and call it if it does. 

Really, I don't think there should be anything special about compilers -- *any* class could have a `runtime_constraints` method.  The only reason it's special for compilers right now is because `runtime_constriants(compiler, pkg)` has to take a compiler spec as a parameter, and that is only because we're in transition from compilers to compilers as proper packages.

it would be nice if in this PR we could get rid of the special casing.  Is there any reason why we can't convert this:

```python
    @classmethod
    def runtime_constraints(cls, compiler, pkg):
        ...
```

to something like:

```python
    @classmethod
    def global_constraints(cls, spec, pkg):
        ...
```

And make the spec an actual *spec* (not compiler spec) for the compiler's associated package? I expect it will need to be this eventually, anyway.  I don't think this is a huge lift, because all we're using the compiler spec for is to get a version and/or name string.  Full fledged specs have those things too.

I think starting with at least that would be good, and eventually I think we want this method to loop over all the possible dependencies and allow them to define global constraints if they need to.  I care about this because I think this pattern has use cases beyond just compilers, and I want the logic to be general.

Reply:
> We shouldn't special case things here.

This specific check is not needed anymore, and will be removed. 

For the API, I think we can inject a spec, but we might need to attach some more information to it. Currently we are not passing a `CompilerSpec`, but a `Compiler`. This means we would be missing information on `cc`, `cxx`, etc. paths (in case they are needed).

Reply:
Opted for passing a `CompilerSpec` for the time being, to avoid double conversions like `oneapi` -> `intel-oneapi-compilers` before the calls, and the opposite within the call.